pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CaseTransformFunction.java

```diff
@@ -0,0 +1,482 @@
+            }
+          }
+          break;
+        case STRING:
+          String[] evalStrings = transformFunction.transformToStringValuesSV(projectionBlock);
+          for (int j = 0; j < selected.length; j++) {
+            if (selected[j] == i) {
+              results[j] = new BigDecimal(evalStrings[i]).floatValue();
+            }
+          }
+          break;
+        default:
+          throw new IllegalStateException(String
+              .format("Cannot convert result type [%s] to [FLOAT] for transform function [%s]", dataType,
+                  transformFunction));
+      }
+    }
+    return results;
+  }
+
+  @Override
+  public double[] transformToDoubleValuesSV(ProjectionBlock projectionBlock) {
+    int[] selected = getSelectedArray(projectionBlock);
+    double[] results = new double[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    for (int i = 0; i < _elseThenStatements.size(); i++) {
+      TransformFunction transformFunction = _elseThenStatements.get(i);
+      FieldSpec.DataType dataType = transformFunction.getResultMetadata().getDataType();
+      switch (dataType) {
+        case INT:
+          int[] evalInts = transformFunction.transformToIntValuesSV(projectionBlock);
+          for (int j = 0; j < selected.length; j++) {
+            if (selected[j] == i) {
+              results[j] = evalInts[j];
+            }
+          }
+          break;
+        case LONG:
+          long[] evalLongs = transformFunction.transformToLongValuesSV(projectionBlock);
+          for (int j = 0; j < selected.length; j++) {
+            if (selected[j] == i) {
+              results[j] = evalLongs[j];
+            }
+          }
+          break;
+        case FLOAT:
+          float[] evalFloats = transformFunction.transformToFloatValuesSV(projectionBlock);
+          for (int j = 0; j < selected.length; j++) {
+            if (selected[j] == i) {
+              results[j] = evalFloats[j];
+            }
+          }
+          break;
+        case DOUBLE:
+          double[] evalDoubles = transformFunction.transformToDoubleValuesSV(projectionBlock);
+          for (int j = 0; j < selected.length; j++) {
+            if (selected[j] == i) {
+              results[j] = evalDoubles[j];
+            }
+          }
+          break;
+        case STRING:
+          String[] evalStrings = transformFunction.transformToStringValuesSV(projectionBlock);
+          for (int j = 0; j < selected.length; j++) {
+            if (selected[j] == i) {
+              results[j] = new BigDecimal(evalStrings[i]).doubleValue();
+            }
+          }
+          break;
+        default:
+          throw new IllegalStateException(String
+              .format("Cannot convert result type [%s] to [DOUBLE] for transform function [%s]", dataType,
+                  transformFunction));
+      }
+    }
+    return results;
+  }
+
+  @Override
+  public String[] transformToStringValuesSV(ProjectionBlock projectionBlock) {
```

Comment:
I am not sure why we need to have types here. If the THEN is a literal expression of type string (the value itself could be numerical), then we are always  going to project string values. Why do we need to run over types here?

Reply:
Right side could also be identifier or transform function. So we cannot do simpler assumptions here.